torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp

```diff
@@ -19,6 +19,12 @@ namespace torch {
 namespace jit {
 namespace fuser {
 
+Statement::Statement(const Statement* src, IrCloner* ir_cloner) {
+  ir_cloner->registerClone(src, this);
```

Comment:
Couldn't registerClone be called inside ir_cloner? Then couldn't we implement `unsafeCloneAs` on all the IR nodes as protected functions? Might just keep IrCloner more isolated from the nodes themselves, although it would require a friend, and I know some people don't like friends.

Reply:
Never mind, realized this wouldn't work later on as clone is protecting against duplicate clones and you're calling clone inside the constructors with the IrCloner. Leaving this here just so you can confirm I understood correctly.

Reply:
Yes, this goes a bit deeper:

(1) registration needs to be the very first thing done after allocating a Statement subclass, otherwise we wouldn't be able to handle cycles in the IR graph (ex. ForLoop::parent_scope_)

(2) in order to split the individual node cloning outside the nodes themselves you need not only friend access but you also need good default constructors for every nodes. It would also require us to remove all the const data members (ex. ForLoop::iter_domain_) since you can't mutate a const after it's constructed. And I'm not sure if it would be a good idea to put something so intimately tied to the members of a class far away from the class itself.

Reply:
okay, makes sense, thanks.